__makeref causes csc.exe exited with code -2146232797

**Version Used**: Visual Studio 16.8.1 | .NET Core 5.0.102

**Steps to Reproduce**:

Code bellow doesn't compile:

```cs
using System;
using System.Threading.Tasks;

static async Task Main(string[] args)
{
    Task<object> task = MethodAsync();
    GetReference(__makeref(task));
    object result = await task; //<-- problem is here
}

static void GetReference(TypedReference reference) { }

static async Task<object> MethodAsync()
{
    return null;
}
```

If we remove variable from return await, works fine:

```cs
///...same code
static async Task Main(string[] args)
{
    Task<object> task = MethodAsync();
    GetReference(__makeref(task));
    await task; //<---- compile normally
}
///...same code
```

**Expected Behavior**:
Successful compile

**Actual Behavior**:
`Microsoft.CSharp.Core.targets(71,5): error MSB6006: "csc.exe" exited with code -2146232797.`
`System.InvalidOperationException: This program location is thought to be unreachable.`

[SharpLap - Error](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYAYAEMEFYDcqqOAbNgBzZkCyAhmAHYAUO6A2gLqa0BOA5gGcAlKgDeqTJKoAeAPYAjAFYBTAMbAAfJmC0BAa0wBeTNWXAAFrIAmAQQEBPBqqZDCKKZgDiZgErKAZso8yk7KTAD64QC2tHpBAUw6+kKuElIKKuqYwQIArhDARtgAnNq6em4AvkRoCGQwACxevgHxoUwAKvYADspWfoHBodmtQ6rKQpiimNUoxHUUMhlqmiZmlrYOTi5iaZIwAOyYDPkQVahAA)

[SharpLab - Working](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYAYAEMEFYDcqqOAbNgBzZkCyAhmAHYAUO6A2gLqa0BOA5gGcAlKgDeqTJKoAeAPYAjAFYBTAMbAAfJmC0BAa0wBeTNWXAAFrIAmAQQEBPBqqZDCKKZgDiZgErKAZso8yk7KTAD64QC2tHpBAUw6+kKuElIwAJzaunpuAL5EaAhkMAAsXr4B8aFMACr2AA7KVn6BwaGYwW0hqspCmKKYBSjExRQyCirqWqYW1naOziIo4u7pAOyYDACuEBD5qEA)

--- a/src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs
+++ b/src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs
@@ -963,7 +963,9 @@ namespace Microsoft.CodeAnalysis.CSharp
 
         public override BoundNode VisitMakeRefOperator(BoundMakeRefOperator node)
         {
-            throw ExceptionUtilities.Unreachable;
+            BoundSpillSequenceBuilder builder = null;
+            var operand = VisitExpression(ref builder, node.Operand);
+            return UpdateExpression(builder, node.Update(operand, node.Type));
         }
 
         public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperator node)

--- a/src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs
+++ b/src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs
@@ -17175,5 +17175,48 @@ class Program
 ";
             var compilation = CompileAndVerify(source, options: TestOptions.ReleaseExe.WithAllowUnsafe(true), verify: Verification.Skipped, expectedOutput: @"");
         }
+
+        [Fact]
+        [WorkItem(51228, "https://github.com/dotnet/roslyn/issues/51228")]
+        public void Issue51228()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+
+class Program
+{
+    static Task<object> t;
+
+    static async Task Main(string[] args)
+    {
+        Task<object> task = MethodAsync();
+        GetReference(__makeref(task));
+        object result = await task;
+        System.Console.WriteLine(result);
+        System.Console.WriteLine(task == t);
+    }
+
+    static void GetReference(TypedReference reference)
+    {
+        t = __refvalue(reference, Task<object>);
+        System.Console.WriteLine(__reftype(reference));
+    }
+
+    static async Task<object> MethodAsync()
+    {
+        await Task.FromResult(1);
+        await Task.FromResult(2);
+        return ""Success"";
+    }
+}
+";
+
+            CompileAndVerify(source, expectedOutput: @"
+System.Threading.Tasks.Task`1[System.Object]
+Success
+True
+").VerifyDiagnostics();
+        }
     }
 }

